### PR TITLE
Gateway refactor

### PR DESCRIPTION
**Pull request by AnturK**
Gateways now have control computers, can handle multiple destinations, are single object.

image

### CODE REVIEW

**Comment by trollbreeder** [comment]
Away missions coming back? Boy, i'm excited!

**Comment by MMMiracles** [comment]
that's hot ngl

**Comment by OverlyGeneric** [comment]
oh boy, I can't wait to see away missions get nerfed to shit or some arbitrary blockade like R&D research, time locks, and recycling telescience maths.

**Comment by Ziiro** [comment]
>Away missions coming back? Boy, i'm excited!

Off station content is being axed, so no

**Comment by AnturK** [comment]
Okay, should be good to review now.

File: code/modules/awaymissions/gateway.dm [hunk 1]
```diff
@@ -1,252 +1,322 @@
...
+	.["ref"] = REF(src)
+	.["name"] = name
+	.["availible"] = is_availible()
+	.["reason"] = get_availible_reason()
+	if(wait)
+		.["timeout"] = max(1 - (wait - world.time) / wait, 0)
+
+/* Destination is another gateway */
+/datum/gateway_destination/gateway
+	var/obj/machinery/gateway/G
```
Comment (spookydonut) [inline code/modules/awaymissions/gateway.dm]: G is a dumb var name, call it like parent or parent_atom or some shit

File: code/modules/awaymissions/gateway.dm [hunk 2]
```diff
@@ -1,252 +1,322 @@
...
+	. = ..()
+	if(!G.calibrated)
+		. = "Exit gateway malfunction. Manual recalibration required."
+	if(G.target)
+		. = "Exit gateway in use."
+	if(!G.powered())
+		. = "Exit gateway unpowered."
+
+/datum/gateway_destination/gateway/get_target_turf()
+	return get_step(G.portal,SOUTH)
```
Comment (spookydonut) [inline code/modules/awaymissions/gateway.dm]: why not have it use the gateway's direction?

File: code/modules/awaymissions/gateway.dm [hunk 3]
```diff
@@ -1,252 +1,322 @@
...
+/datum/gateway_destination/point/get_target_turf()
+	return pick(target_turfs)
+
+/* Dense invisible object starting the teleportation. Created by gateways on activation. */
+/obj/effect/gateway_portal_bumper
+	var/obj/machinery/gateway/gateway
+	density = TRUE
+	invisibility = INVISIBILITY_ABSTRACT
+
+/obj/effect/gateway_portal_bumper/Bumped(atom/movable/AM)
```
Comment (spookydonut) [inline code/modules/awaymissions/gateway.dm]: make it optionally directional where it can only be entered from a specific direction?

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/modules/awaymissions/gateway.dm (hunk 1 above)
Comment (spookydonut) [reply code/modules/awaymissions/gateway.dm]: also autodoc

**Comment by stylemistake** [comment]
I'd merge if all other reviews were addressed.

**Pull request merged**